File: api/share.py
"""
share.py — Compartir paquetes de palabras entre usuarios
"""
import json
import secrets
from datetime import date
from flask import Blueprint, request, g
from api.config import (
    ok, err, body, db_exec, db_fetchall, db_fetchone,
    db_insert, db_update, APP_URL, require_auth
)

share_bp = Blueprint("share", __name__)


@share_bp.route("/share", methods=["GET", "POST", "DELETE", "OPTIONS"])
def share():
    if request.method == "OPTIONS":
        return "", 204

    method = request.method
    action = request.args.get("action", "")

    # ── Public endpoints (no auth required) ─────────────────────────────────

    # GET ?action=get&token=XXX
    if method == "GET" and action == "get":
        token = request.args.get("token", "").strip()
        if not token:
            return err("Token requerido")
        pack = db_fetchone("SELECT * FROM shared_packs WHERE token = %s", (token,))
        if not pack:
            return err("Paquete no encontrado o expirado", 404)
        owner = db_fetchone("SELECT username FROM users WHERE id = %s", (pack["user_id"],))
        return ok({
            "token": pack["token"],
            "label": pack["label"],
            "word_count": int(pack["word_count"]),
            "owner": owner["username"] if owner else "usuario",
            "import_count": int(pack["import_count"]),
            "created_at": str(pack["created_at"]),
            "words": json.loads(pack["words_json"]),
        })

    # GET ?action=public_packs
    if method == "GET" and action == "public_packs":
        cat = request.args.get("category", "").strip()
        if cat:
            packs = db_fetchall(
                """SELECT sp.token, sp.label, COALESCE(sp.category,'') AS category,
                          sp.word_count, sp.import_count, u.username AS owner
                   FROM shared_packs sp JOIN users u ON u.id = sp.user_id
                   WHERE COALESCE(sp.is_public,0) = 1 AND sp.category = %s
                   ORDER BY sp.import_count DESC, sp.created_at DESC LIMIT 50""",
                (cat,),
            )
        else:
            packs = db_fetchall(
                """SELECT sp.token, sp.label, COALESCE(sp.category,'') AS category,
                          sp.word_count, sp.import_count, u.username AS owner
                   FROM shared_packs sp JOIN users u ON u.id = sp.user_id
                   WHERE COALESCE(sp.is_public,0) = 1
                   ORDER BY sp.import_count DESC, sp.created_at DESC LIMIT 50""",
            )
        try:
            cats = [r["category"] for r in db_fetchall(
                "SELECT DISTINCT category FROM shared_packs WHERE COALESCE(is_public,0)=1 AND category != '' ORDER BY category"
            )]
        except Exception:
            cats = []
        return ok({"packs": packs, "categories": cats})

    # ── Auth-required endpoints ──────────────────────────────────────────────
    from api.config import get_uid
    uid = get_uid()
    if uid is None and action in ("create", "import", "mine", "delete"):
        return err("No autenticado", 401)

    # POST ?action=create
    if method == "POST" and action == "create":
        b = body()
        label = b.get("label", "").strip()
        category = b.get("category", "").strip()
        is_public = int(bool(b.get("is_public", False)))
        group_ids = b.get("group_ids", [])
        date_val = b.get("date")
        dates = b.get("dates", [])
        words_direct = b.get("_words_direct", [])

        if date_val:
            dates = [date_val]

        words = []
        if words_direct:
            for w in words_direct:
                sp = w.get("spanish", "").strip()
                en = [e.strip() for e in (w.get("english") or []) if str(e).strip()]
                if sp and en:
                    words.append({"spanish": sp, "english": en})
        elif group_ids:
            ph = ",".join(["%s"] * len(group_ids))
            rows = db_fetchall(
                f"""SELECT g.id, g.spanish,
                           GROUP_CONCAT(w.english ORDER BY w.id SEPARATOR '||') AS english_words
                    FROM word_groups g JOIN words w ON w.group_id = g.id
                    WHERE g.user_id = %s AND g.id IN ({ph})
                    GROUP BY g.id""",
                tuple([uid] + group_ids),
            )
            for r in rows:
                words.append({"spanish": r["spanish"], "english": r["english_words"].split("||")})
        elif dates:
            ph = ",".join(["%s"] * len(dates))
            rows = db_fetchall(
                f"""SELECT g.id, g.spanish,
                           GROUP_CONCAT(w.english ORDER BY w.id SEPARATOR '||') AS english_words
                    FROM word_groups g JOIN words w ON w.group_id = g.id
                    WHERE g.user_id = %s AND DATE(g.created_at) IN ({ph})
                    GROUP BY g.id""",
                tuple([uid] + dates),
            )
            for r in rows:
                words.append({"spanish": r["spanish"], "english": r["english_words"].split("||")})
        else:
            return err("Debes especificar group_ids, date, dates o _words_direct")

        if not words:
            return err("No hay palabras para compartir")
        if not label:
            label = f"{len(words)} palabras"

        token = secrets.token_hex(16)
        db_insert(
            "INSERT INTO shared_packs (token, user_id, label, category, is_public, words_json, word_count) VALUES (%s, %s, %s, %s, %s, %s, %s)",
            (token, uid, label, category, is_public, json.dumps(words, ensure_ascii=False), len(words)),
        )
        return ok({
            "token": token,
            "word_count": len(words),
            "label": label,
            "category": category,
            "is_public": bool(is_public),
            "url": f"{APP_URL}/?share={token}",
        })

    # POST ?action=import&token=XXX
    if method == "POST" and action == "import":
        token = request.args.get("token", "").strip()
        if not token:
            return err("Token requerido")
        pack = db_fetchone("SELECT * FROM shared_packs WHERE token = %s", (token,))
        if not pack:
            return err("Paquete no encontrado", 404)

        words = json.loads(pack["words_json"]) or []
        added = 0
        skipped = 0
        group_ids = []

        from api.config import get_db
        conn = get_db()
        conn.begin()
        try:
            for w in words:
                spanish = w.get("spanish", "").strip()
                english = [e.strip() for e in w.get("english", []) if str(e).strip()]
                if not spanish or not english:
                    skipped += 1
                    continue
                dup = db_fetchone(
                    "SELECT id FROM word_groups WHERE user_id = %s AND spanish = %s LIMIT 1",
                    (uid, spanish),
                )
                if dup:
                    skipped += 1
                    continue
                with conn.cursor() as cur:
                    cur.execute(
                        "INSERT INTO word_groups (user_id, spanish, created_at) VALUES (%s, %s, NOW())",
                        (uid, spanish),
                    )
                    gid = cur.lastrowid
                    group_ids.append(int(gid))
                    for en in english:
                        cur.execute(
                            "INSERT INTO words (group_id, english, is_hard) VALUES (%s, %s, 0)",
                            (gid, en),
                        )
                added += 1
            conn.commit()
            db_update("UPDATE shared_packs SET import_count = import_count + 1 WHERE token = %s", (token,))
        except Exception as e:
            conn.rollback()
            return err(f"Error al importar: {e}", 500)

        return ok({"added": added, "skipped": skipped, "group_ids": group_ids})

    # GET ?action=mine
    if method == "GET" and action == "mine":
        rows = db_fetchall(
            """SELECT token, label, COALESCE(category,'') AS category,
                      COALESCE(is_public,0) AS is_public, word_count, import_count, created_at
               FROM shared_packs WHERE user_id = %s ORDER BY created_at DESC LIMIT 20""",
            (uid,),
        )
        for r in rows:
            if r.get("created_at"):
                r["created_at"] = str(r["created_at"])
        return ok(rows)

    # DELETE ?action=delete&token=XXX
    if method == "DELETE" and action == "delete":
        token = request.args.get("token", "").strip()
        if not token:
            return err("Token requerido")
        affected = db_update(
            "DELETE FROM shared_packs WHERE token = %s AND user_id = %s", (token, uid)
        )
        if affected == 0:
            return err("Paquete no encontrado", 404)
        return ok(None)

    return err("Acción no válida", 400)

File: api/share.php
<?php
// ============================================================
//  share.php — Compartir paquetes de palabras entre usuarios
// ============================================================
require __DIR__ . '/config.php';

// ── Migrar tabla de paquetes compartidos si no existe ───────
try {
    db()->exec("
        CREATE TABLE IF NOT EXISTS shared_packs (
            id          INT AUTO_INCREMENT PRIMARY KEY,
            token       VARCHAR(32) NOT NULL UNIQUE,
            user_id     INT NOT NULL,
            label       VARCHAR(120) NOT NULL DEFAULT '',
            category    VARCHAR(80)  NOT NULL DEFAULT '',
            is_public   TINYINT(1)   NOT NULL DEFAULT 0,
            words_json  MEDIUMTEXT NOT NULL,
            word_count  INT NOT NULL DEFAULT 0,
            created_at  TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
            expires_at  DATETIME NULL DEFAULT NULL,
            import_count INT NOT NULL DEFAULT 0
        ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4
    ");
} catch (Throwable $e) {}
// Migrar columnas si ya existía la tabla
try { db()->exec("ALTER TABLE shared_packs ADD COLUMN category VARCHAR(80) NOT NULL DEFAULT ''"); } catch(Throwable $e) {}
try { db()->exec("ALTER TABLE shared_packs ADD COLUMN is_public TINYINT(1) NOT NULL DEFAULT 0"); } catch(Throwable $e) {}

try { db()->exec("ALTER TABLE shared_packs ADD COLUMN words_json MEDIUMTEXT NOT NULL DEFAULT ''"); } catch(Throwable $e) {}
try { db()->exec("ALTER TABLE shared_packs ADD COLUMN word_count INT NOT NULL DEFAULT 0"); } catch(Throwable $e) {}
try { db()->exec("ALTER TABLE shared_packs ADD COLUMN import_count INT NOT NULL DEFAULT 0"); } catch(Throwable $e) {}
try { db()->exec("ALTER TABLE shared_packs ADD COLUMN expires_at DATETIME NULL DEFAULT NULL"); } catch(Throwable $e) {}
try { db()->exec("ALTER TABLE shared_packs ADD COLUMN label VARCHAR(120) NOT NULL DEFAULT ''"); } catch(Throwable $e) {}

// Capturar errores fatales y retornar JSON en vez de HTML
set_exception_handler(function(Throwable $e) {
    http_response_code(500);
    echo json_encode(['ok' => false, 'error' => $e->getMessage()]);
    exit;
});

$method = $_SERVER['REQUEST_METHOD'];
$action = $_GET['action'] ?? '';

// ════════════════════════════════════════════════════════════
//  POST ?action=create — Crear un paquete compartido
//  Requiere auth. Body: { label, group_ids[] | date | dates[] }
// ════════════════════════════════════════════════════════════
if ($method === 'POST' && $action === 'create') {
    $uid  = uid();
    $b    = body();
    $label     = trim($b['label']    ?? '');
    $category  = trim($b['category']  ?? '');
    $is_public = (int)(bool)($b['is_public'] ?? false);
    $pdo  = db();

    // Obtener los grupos a compartir
    $groupIds = $b['group_ids'] ?? [];
    $date     = $b['date'] ?? null;
    $dates    = $b['dates'] ?? [];

    if ($date) $dates = [$date];

    $words = [];

    $wordsDirect = $b['_words_direct'] ?? [];

    if (!empty($wordsDirect)) {
        // Palabras pasadas directamente (desde importación JSON)
        foreach ($wordsDirect as $w) {
            $sp = trim($w['spanish'] ?? '');
            $en = array_values(array_filter(array_map('trim', (array)($w['english'] ?? []))));
            if ($sp && $en) $words[] = ['spanish' => $sp, 'english' => $en];
        }
    } elseif (!empty($groupIds)) {
        // Por IDs específicos
        $ph   = implode(',', array_fill(0, count($groupIds), '?'));
        $stmt = $pdo->prepare("
            SELECT g.id, g.spanish,
                   GROUP_CONCAT(w.english ORDER BY w.id SEPARATOR '||') AS english_words,
                   GROUP_CONCAT(w.is_hard ORDER BY w.id SEPARATOR '||') AS english_diffs
            FROM word_groups g
            JOIN words w ON w.group_id = g.id
            WHERE g.user_id = ? AND g.id IN ($ph)
            GROUP BY g.id
        ");
        $stmt->execute(array_merge([$uid], $groupIds));
        $rows = $stmt->fetchAll();
        foreach ($rows as $r) {
            $words[] = ['spanish' => $r['spanish'], 'english' => explode('||', $r['english_words'])];
        }
    } elseif (!empty($dates)) {
        // Por fecha(s)
        $ph   = implode(',', array_fill(0, count($dates), '?'));
        $stmt = $pdo->prepare("
            SELECT g.id, g.spanish,
                   GROUP_CONCAT(w.english ORDER BY w.id SEPARATOR '||') AS english_words,
                   GROUP_CONCAT(w.is_hard ORDER BY w.id SEPARATOR '||') AS english_diffs
            FROM word_groups g
            JOIN words w ON w.group_id = g.id
            WHERE g.user_id = ? AND DATE(g.created_at) IN ($ph)
            GROUP BY g.id
        ");
        $stmt->execute(array_merge([$uid], $dates));
        $rows = $stmt->fetchAll();
        foreach ($rows as $r) {
            $words[] = ['spanish' => $r['spanish'], 'english' => explode('||', $r['english_words'])];
        }
    } else {
        err('Debes especificar group_ids, date, dates o _words_direct');
    }

    if (!$words) err('No hay palabras para compartir');

    if (!$label) $label = count($words) . ' palabras';

    $token = bin2hex(random_bytes(16));
    $pdo->prepare("
        INSERT INTO shared_packs (token, user_id, label, category, is_public, words_json, word_count)
        VALUES (?, ?, ?, ?, ?, ?, ?)
    ")->execute([$token, $uid, $label, $category, $is_public, json_encode($words), count($words)]);

    ok(['token' => $token, 'word_count' => count($words), 'label' => $label, 'category' => $category, 'is_public' => (bool)$is_public,
        'url' => APP_URL . '/?share=' . $token]);
}

// ════════════════════════════════════════════════════════════
//  GET ?action=get&token=XXX — Ver un paquete (sin auth)
// ════════════════════════════════════════════════════════════
if ($method === 'GET' && $action === 'get') {
    $token = trim($_GET['token'] ?? '');
    if (!$token) err('Token requerido');

    $stmt = db()->prepare("SELECT * FROM shared_packs WHERE token = ?");
    $stmt->execute([$token]);
    $pack = $stmt->fetch();
    if (!$pack) err('Paquete no encontrado o expirado', 404);

    // Obtener username del dueño
    $uStmt = db()->prepare("SELECT username FROM users WHERE id = ?");
    $uStmt->execute([$pack['user_id']]);
    $owner = $uStmt->fetch();

    ok([
        'token'        => $pack['token'],
        'label'        => $pack['label'],
        'word_count'   => (int)$pack['word_count'],
        'owner'        => $owner['username'] ?? 'usuario',
        'import_count' => (int)$pack['import_count'],
        'created_at'   => $pack['created_at'],
        'words'        => json_decode($pack['words_json'], true),
    ]);
}

// ════════════════════════════════════════════════════════════
//  POST ?action=import&token=XXX — Importar paquete a mi cuenta
// ════════════════════════════════════════════════════════════
if ($method === 'POST' && $action === 'import') {
    $uid   = uid();
    $token = trim($_GET['token'] ?? '');
    if (!$token) err('Token requerido');

    $pdo  = db();
    $stmt = $pdo->prepare("SELECT * FROM shared_packs WHERE token = ?");
    $stmt->execute([$token]);
    $pack = $stmt->fetch();
    if (!$pack) err('Paquete no encontrado', 404);

    $words  = json_decode($pack['words_json'], true) ?? [];
    $added  = 0;
    $skipped = 0;
    $groupIds = [];

    $pdo->beginTransaction();
    try {
        foreach ($words as $w) {
            $spanish = trim($w['spanish'] ?? '');
            $english = array_filter(array_map('trim', $w['english'] ?? []));
            if (!$spanish || !$english) { $skipped++; continue; }

            // Check dup
            $dup = $pdo->prepare("SELECT id FROM word_groups WHERE user_id = ? AND spanish = ? LIMIT 1");
            $dup->execute([$uid, $spanish]);
            if ($dup->fetch()) { $skipped++; continue; }

            $ins = $pdo->prepare("INSERT INTO word_groups (user_id, spanish, created_at) VALUES (?, ?, NOW())");
            $ins->execute([$uid, $spanish]);
            $gid = $pdo->lastInsertId();
            $groupIds[] = (int)$gid;

            $insW = $pdo->prepare("INSERT INTO words (group_id, english, is_hard) VALUES (?, ?, 0)");
            foreach ($english as $en) $insW->execute([$gid, $en]);

            $added++;
        }
        $pdo->commit();
        // Increment import counter
        $pdo->prepare("UPDATE shared_packs SET import_count = import_count + 1 WHERE token = ?")->execute([$token]);
    } catch (Throwable $e) {
        $pdo->rollBack();
        err('Error al importar: ' . $e->getMessage(), 500);
    }

    ok(['added' => $added, 'skipped' => $skipped, 'group_ids' => $groupIds]);
}

// ════════════════════════════════════════════════════════════
//  GET ?action=public_packs[&category=X] — Packs públicos (sin auth)
// ════════════════════════════════════════════════════════════
if ($method === 'GET' && $action === 'public_packs') {
    // Asegurar columnas existen antes de consultarlas
    try { db()->exec("ALTER TABLE shared_packs ADD COLUMN category VARCHAR(80) NOT NULL DEFAULT ''"); } catch(Throwable $e) {}
    try { db()->exec("ALTER TABLE shared_packs ADD COLUMN is_public TINYINT(1) NOT NULL DEFAULT 0"); } catch(Throwable $e) {}

    $cat    = trim($_GET['category'] ?? '');
    $where  = $cat ? "WHERE COALESCE(sp.is_public,0) = 1 AND sp.category = ?" : "WHERE COALESCE(sp.is_public,0) = 1";
    $params = $cat ? [$cat] : [];

    try {
        $stmt = db()->prepare("
            SELECT sp.token, sp.label, COALESCE(sp.category,'') AS category,
                   sp.word_count, sp.import_count, u.username AS owner
            FROM shared_packs sp
            JOIN users u ON u.id = sp.user_id
            $where
            ORDER BY sp.import_count DESC, sp.created_at DESC
            LIMIT 50
        ");
        $stmt->execute($params);
        $packs = $stmt->fetchAll();
        $cats = db()->query("SELECT DISTINCT category FROM shared_packs WHERE COALESCE(is_public,0)=1 AND category != '' ORDER BY category")->fetchAll(PDO::FETCH_COLUMN);
    } catch (Throwable $e) {
        $packs = []; $cats = [];
        // Debug: log error (remove after fix)
        error_log('share.php public_packs error: ' . $e->getMessage());
    }

    ok(['packs' => $packs, 'categories' => $cats]);
}

// ════════════════════════════════════════════════════════════
//  GET ?action=mine — Mis paquetes compartidos
// ════════════════════════════════════════════════════════════
if ($method === 'GET' && $action === 'mine') {
    $uid  = uid();
    try {
        $stmt = db()->prepare("
            SELECT token, label, COALESCE(category,'') AS category,
                   COALESCE(is_public,0) AS is_public, word_count, import_count, created_at
            FROM shared_packs
            WHERE user_id = ?
            ORDER BY created_at DESC
            LIMIT 20
        ");
        $stmt->execute([$uid]);
        ok($stmt->fetchAll());
    } catch (Throwable $e) {
        // Columnas nuevas no existen aún — fallback sin ellas
        $stmt = db()->prepare("SELECT token, label, word_count, import_count, created_at FROM shared_packs WHERE user_id = ? ORDER BY created_at DESC LIMIT 20");
        $stmt->execute([$uid]);
        $rows = $stmt->fetchAll();
        foreach ($rows as &$r) { $r['category'] = ''; $r['is_public'] = false; }
        ok($rows);
    }
}

// ════════════════════════════════════════════════════════════
//  DELETE ?action=delete&token=XXX — Borrar un paquete
// ════════════════════════════════════════════════════════════
if ($method === 'DELETE' && $action === 'delete') {
    $uid   = uid();
    $token = trim($_GET['token'] ?? '');
    if (!$token) err('Token requerido');
    $stmt = db()->prepare("DELETE FROM shared_packs WHERE token = ? AND user_id = ?");
    $stmt->execute([$token, $uid]);
    if ($stmt->rowCount() === 0) err('Paquete no encontrado', 404);
    ok(null);
}

err('Acción no válida', 400);